app/ece.js

```diff
@@ -0,0 +1,237 @@
+require('buffer');
+
+const NONCE_LENGTH = 12;
+const TAG_LENGTH = 16;
+const KEY_LENGTH = 16;
+const MODE_ENCRYPT = 'encrypt';
+const MODE_DECRYPT = 'decrypt';
+
+const encoder = new TextEncoder();
+
+function generateSalt(len) {
+  const randSalt = new Uint8Array(len);
+  window.crypto.getRandomValues(randSalt);
+  return randSalt.buffer;
+}
+
+/*
+mode: string, either 'encrypt' or 'decrypt'
+ikm:  Uint8Array containing key of KEY_LENGTH length
+rs:   int containing record size, optional
+salt: ArrayBuffer containing salt of KEY_LENGTH length, optional
+The transform stream takes data as UInt8Arrays on the writable side, and outputs
+UInt8Arrays on the readable side.
+*/
+export default class ECETransformer {
+  constructor(mode, ikm, rs, salt) {
+    this.mode = mode;
+    this.prevChunk;
+    this.params = {};
+    this.seq = 0;
+    this.firstchunk = true;
+    this.rs = rs || 1024;
+    this.ikm = ikm.buffer;
+    this.params.salt = salt;
+    if (!salt) {
+      this.params.salt = generateSalt(KEY_LENGTH);
+    }
+  }
+
+  async generateKey() {
+    const inputKey = await window.crypto.subtle.importKey(
+      'raw',
+      this.ikm,
+      'HKDF',
+      false,
+      ['deriveKey']
+    );
+
+    return window.crypto.subtle.deriveKey(
+      {
+        name: 'HKDF',
+        salt: this.params.salt,
+        info: encoder.encode('Content-Encoding: aes128gcm\0'),
+        hash: 'SHA-256'
+      },
+      inputKey,
+      {
+        name: 'AES-GCM',
+        length: 128
+      },
+      false,
+      ['encrypt', 'decrypt']
+    );
+  }
+
+  async generateNonceBase() {
+    const inputKey = await window.crypto.subtle.importKey(
+      'raw',
+      this.ikm,
+      'HKDF',
+      false,
+      ['deriveKey']
+    );
+
+    const base = await window.crypto.subtle.exportKey(
+      'raw',
+      await window.crypto.subtle.deriveKey(
+        {
+          name: 'HKDF',
+          salt: this.params.salt,
+          info: encoder.encode('Content-Encoding: nonce\0'),
+          hash: 'SHA-256'
+        },
+        inputKey,
+        {
+          name: 'AES-GCM',
+          length: 128
+        },
+        true,
+        ['encrypt', 'decrypt']
+      )
+    );
+
+    return Buffer.from(base.slice(0, NONCE_LENGTH));
+  }
+
+  generateNonce(seq) {
+    const nonce = Buffer.from(this.params.nonceBase);
+    const m = nonce.readUIntBE(nonce.length - 4, 4);
+    const xor = (m ^ seq) >>> 0; //forces unsigned int xor
+    nonce.writeUIntBE(xor, nonce.length - 4, 4);
+
+    const m2 = nonce.readUIntBE(nonce.length - 8, 4);
+    const xor2 = (m2 ^ (seq >>> 4)) >>> 0;
```

Comment:
IMPORTANT: you want to use a shift by 32, but that isn't going to work (bitwise ops in JS operate on 32-bit values).  Using floating point ops is necessary here, or you could decide not to support more than 2^32 records.